File: Sort/merge_intervals.py
class Solution(object):
    def merge(self, intervals):
        """
        :type intervals: List[List[int]]
        :rtype: List[List[int]]
        给定一个包含一些代表区间的数组，要求合并有重叠的数组项，边界相等也算重叠
        这里的重叠存在两种情况：1.相交 2.包含
        首先，先将区间按照右边界点排序，为什么是右边界点，为了处理有某个大区间包含了前面很多个小区间的情况
        然后从后面往前面遍历，按照上面的两种情况看有没有重叠，
        如果有则合并，[min(left), max(right)]
        """
        if len(intervals) < 2:
            return intervals

        # sorted 返回的是副本，不改变原来的元素
        intervals = sorted(intervals, key = lambda x: x[1])

        for i in range(len(intervals)-1, 0, -1):
            if intervals[i][0] <= intervals[i-1][1]:
                intervals[i-1][0] = min(intervals[i-1][0], intervals[i][0])
                intervals[i-1][1] = intervals[i][1]     # 右边界是排过序的

            intervals.pop(i)

        return intervals

File: Sort/MergeIntervals.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;

class MergeIntervals {
    public int[][] merge(int[][] intervals) {
        if (intervals == null || intervals.length == 0) return intervals;

        List<int[]> result = new ArrayList<>();

        // 首先按每个区间的第一个元素进行排序
        Arrays.sort(intervals, new Comparator<int[]>() {
            public int compare(int[] o1, int[] o2) {
                if (o1[0] < o2[0]) {
                    return -1;
                } else if (o1[0] > o2[0]) {
                    return 1;
                } else {
                    return 0;
                }
            }
        });

        result.add(intervals[0]);

        // 排序后的区间只要检查它和上一个区间是否有重叠，
        // 如果重叠了就求出合并后的区间，加到结果集中
        // 如果没有则直接加到结果集中
        for (int i = 1; i < intervals.length; i++) {
            if (result.get(result.size() - 1)[1] >= intervals[i][0]) {
                // 有可能第一个区间包含了第二个区间
                int[] temp = {result.get(result.size()-1)[0], Math.max(result.get(result.size() - 1)[1], intervals[i][1])};
                result.remove(result.size() - 1);
                result.add(temp);
            } else {
                result.add(intervals[i]);
            }
        }

        int[][] result2Int = new int[result.size()][];

        for (int i = 0; i < result.size(); i++) {
            result2Int[i] = result.get(i);
        }

        return result2Int;



    }

    public static void main(String[] args) {
        int[][] nums = {{1, 3}, {8, 10}, {2, 6}, {15, 18}};
        MergeIntervals solution = new MergeIntervals();
        int[][] result = solution.merge(nums);
        for (int[] num : result) {
            System.out.println(Arrays.toString(num));
        }
    }
}